File: 5.term BIT/IW5 - Programming .NET and C#/Tournament.API.App/Controllers/BaseController.cs
using Microsoft.AspNetCore.Mvc;
using Tournament.API.BL.Facades.Interfaces;
using Tournament.API.BL.Models.DetailModels.Interfaces;
using Tournament.API.BL.Models.FormDefaultModels.Interfaces;
using Tournament.API.BL.Models.FormModels.Interfaces;
using Tournament.API.BL.Models.ListModels.Interfaces;

namespace Tournament.API.App.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class BaseController<TDetailModel, TListModel, TFormModel, TFormDefaultModel> : ControllerBase
    where TDetailModel : class, IDetailModel
    where TListModel : class, IListModel
    where TFormModel : class, IFormModel
    where TFormDefaultModel : class, IFormDefaultModel
    {
        protected readonly IFacadeBase<TDetailModel, TListModel, TFormModel, TFormDefaultModel> Facade;

        public BaseController(IFacadeBase<TDetailModel, TListModel, TFormModel, TFormDefaultModel> facade)
        {
            Facade = facade;
        }

        [HttpGet]
        public ActionResult<IList<TListModel>> GetAll()
        {
            return new ActionResult<IList<TListModel>>(Facade.GetAll());
        }

        [HttpGet("GetFormDefaultModel")]
        public ActionResult<TFormDefaultModel> GetFormDefaultModel()
        {
            return new ActionResult<TFormDefaultModel>(Facade.GetDefaultFormModel());
        }

        [HttpGet("GetById")]
        public ActionResult<TDetailModel> GetById(Guid id)
        {
            var model = Facade.GetById(id);

            if (model == null)
            {
                return NotFound();
            }

            return model;
        }

        [HttpPost]
        public ActionResult<TListModel> Create([FromForm] TFormModel formModel)
        {
            return Facade.Add(formModel);
        }

        [HttpPut]
        public ActionResult<TListModel> Update([FromForm] TFormModel formModel)
        {
            return Facade.Update(formModel);
        }

        [HttpDelete]
        public IActionResult Delete(Guid id)
        {
            Facade.Remove(id);
            return Ok();
        }
    }
}

File: BIT - Bachelor IT/5.term/IIS - Information Systems/serverAPI/app/Http/Controllers/BaseController.php
<?php

/**
* Author: Pavel Šesták
*/

namespace App\Http\Controllers;

use App\Models\Conference;
use Illuminate\Http\Request;

/**
 * @OA\Info(title="eConference API", version="1.0")
 */
class BaseController extends Controller
{
    protected $model;

    public function __construct($model)
    {
        $this->model = $model;
    }

    public function getAll()
    {
        return $this->model::all();
    }

    public function getById(Request $request)
    {
        $id = $request->id;
        $model = $this->model::findOrFail($id);
        return $model;
    }

    public function create(Request $request)
    {
        $this->model::create($request->all());
        return $this->model;
    }

    public function update(Request $request)
    {
        $id = $request->id;
        $model = $this->model::findOrFail($id);
        $model->update($request->all());
        return $model;
    }

    public function delete(Request $request)
    {
        $id = $request->id;
        $model = $this->model::findOrFail($id);
        return $model->delete();
    }
}
